docs/design/sql/06-plan-caching.md

```diff
@@ -0,0 +1,129 @@
+On the contrary, the same query string may resolve different objects for different search paths, as shown in section 1.2. 
+Therefore, search paths must be part of the key.  
+
+### 2.2 Data Structure
+
+We use `ConcurrentHashMap` to store cached plans. `PlanKey` is a key, the plan is a value.
+ 
+### 2.3 Maximum Size and Plan Eviction
+
+The maximum size of the cache is required to prevent out-of-memory if too many distinct queries are submitted.
+
+When a plan is added to the map, the map size is checked. If the map size is greater than the maximum size, some plans are 
+evicted. 
+
+Eviction is synchronous because the asynchronous variant is prone to out-of-memory. We assume that for the most workloads 
+evictions should be rare. 
+
+We use the LRU (least recently used) approach to find the plans to evict. Whenever a plan is accessed, it's `lastUsed` field is 
+updated with the current time. During the eviction, plans are sorted by their `lastUsed` values, and the least recently used
+plans are removed.   
+
+### 2.4 Reacting to Catalog and Partition Distribution Changes
+
+If a catalog or partition distribution is changed, some plans must be invalidated. There are two different ways to achieve 
+this: `push` and `pull`.
+
+With the `push` approach, the plan cache is notified about a change, from the relevant component. E.g., if an index is created,
+then the map service notifies the plan cache about the change. The advantage of this solution is that any change is reflected 
+in the plan cache immediately. However, this approach increases coupling, because many components (map service, 
+replicated map service, partition service) now have to be aware of the SQL subsystem. This approach also requires complex
+synchronization between query optimizer, plan cache, and dependent components, to ensure that no stale plan is ever cached.
+
+With the `pull` approach, the SQL subsystem queries other components periodically, collects the changes, and invalidates 
+affected plans. The advantage of this approach is simplicity. No synchronization or changes to other components are needed. 
+Invalid plans are guaranteed to be removed eventually. The downside is that invalid plans might be active for some time after
+the change has occurred.
+
+We choose the `pull` approach due to simplicity and sufficient guarantees. The background worker reconstructs the catalog 
+periodically, and verifies that existing plans are compatible with the current catalog and partition distribution. 
+
+To counter the problem with outdated plans, we add a special `invalidatePlan` flag to `QueryException`. If an invalid plan
+is used, an exception with this flag will be thrown at some point. When the initiator member receives an exception with 
+this flag, the plan is invalidated. 
+
+Note that currently, users will have to re-execute the query in this case. In future versions, we will add a transparent
+query retry, so that invalid plans will not be visible to users.    
```

Comment:
Is it going to be documented anywhere? How the users can detect that the exception indicates an invalid plan?

Reply:
This will be implemented in a way transparent for users. No documentation is needed.

Reply:
Not sure that I got it. According to this document, the users will have to re-execute the query. Is it the behavior in 4.1? If so, it should be documented.

Reply:
I am not sure where exactly you propose to document it?

Reply:
In HZ manual

Reply:
We will document error conditions in the manual, this is out of scope if this PR. Note that there is no such thing as “plan” or “plan invalidation” from the user perspective. Instead, there are some error conditions, such as map destroy or partition migration, which may lead to plan invalidation, which is purely internal thing. The improvement I mentioned (retries on plan invalidation), will help us hide such errors from users sometimes, but not always.

Reply:
We will document error conditions in the manual, this is out of scope if this PR. Note that there is no such thing as “plan” or “plan invalidation” from the user perspective. Instead, there are some error conditions, such as map destroy or partition migration, which may lead to plan invalidation, which is purely internal thing. The improvement I mentioned (retries on plan invalidation), will help us hide such errors from users sometimes, but not always.